### ISSUE TITLE
Add option to trigger sync hooks on partial syncs

### ISSUE THREAD

**Issue by praneethkaturi** [argoproj/argo-cd#16696]
# Summary

The proposed change aims to extend the current capabilities of Argo CD by introducing an option that allows sync hooks to be triggered during partial or selective sync operations. Currently, sync hooks in Argo CD are designed to execute during full sync processes. However, there are use cases, particularly in complex deployment environments, where partial or selective syncs are frequently used. In these scenarios, the inability to trigger sync hooks can lead to gaps in important operational procedures such as auditing, compliance checks, or custom validation and testing routines.

# Motivation

 In scenarios where every sync in Argo CD is monitored for auditing purposes, and additional checks (like pre-validation RBAC or post-validation testing) are required, the inability to trigger hooks during selective/partial syncs can lead to bypassing these critical steps. Implementing an option to enforce these hooks for all types of syncs would enhance the control and security of the sync process.

# Proposal

Introduce an annotation within the Argo CD application or alongside the sync hook. This annotation would indicate whether triggering hooks for all types of syncs (including partial/selective) is required. This feature would allow users to specify their need for comprehensive hook execution, thereby ensuring compliance with audit and validation procedures.

**Comment by praneethkaturi** [comment]
I'm eager to discuss this feature further with the community. If deemed valuable, I would be excited to contribute to its development.

**Comment by todaywasawesome** [comment]
> I'm eager to discuss this feature further with the community. If deemed valuable, I would be excited to contribute to its development.

Do you want to bring this to the next contributor's [experience meeting on Dec 28th](https://docs.google.com/document/d/1xkoFkVviB70YBzSEa4bDnu-rUZ1sIFtwKKG1Uw8XsY8/edit)?

**Comment by praneethkaturi** [comment]
> Do you want to bring this to the next contributor's [experience meeting on Dec 28th](https://docs.google.com/document/d/1xkoFkVviB70YBzSEa4bDnu-rUZ1sIFtwKKG1Uw8XsY8/edit)?

Certainly, I'm excited to discuss this in our meeting. I've included it in the document. Since this is my first time suggesting a feature, I'd really appreciate your guidance on the preparations needed prior to the meeting.

**Comment by praneethkaturi** [comment]
@todaywasawesome Hey Dan! Looks like we didn't have a lot of attendance this time around, I'll bring it up in the next meeting. Meanwhile, could you please let me know if it is just an open discussion or is any prep needed from my side to present the feature?

**Comment by alexmt** [comment]
The issue was discussed during the contributors' meeting. We've come up with two proposals:

* Implement a new sync option that would allow users to request running hooks even if sync is partial. All sync options can be specified at the app level as a default. Proposal would not be required and next step it to just implement it ( cc @praneethkaturi  in case you want to prepare PR  )
* we've agreed that it makes sense to have separate RBAC check for the partial sync. The use case is to allow administrators to disable partial syncs. The main concern here is to make behavior backward compatible. Separate issue is required for this change.

**Comment by praneethkaturi** [comment]
@alexmt  Thank you for summarizing it, yes I will start working on the PR!

**Comment by pradithya** [comment]
@praneethkaturi Is there any work on-going to support this feature?

**Comment by Christian-Oleson** [comment]
I have opened the [PR](https://github.com/argoproj/argo-cd/pull/26044), though I admittedly did not get this issue assigned to me yet. Please do so if/when possible. Thanks!
